File: server/common/deserialize_utils.py
import struct


# https://docs.python.org/3/library/struct.html#format-characters
def deserialize_bool(buffer) -> bool:
    """
    Unpacks buffer into a boolean
    """
    return struct.unpack('?', buffer)[0]

def deserialize_u8(buffer):
    """
    Unpacks buffer into a uint8
    """
    return int(struct.unpack("!B", buffer)[0])

def deserialize_u16(buffer):
    """
    Unpacks buffer into a uint16 from network order (big endian into the machine endianness)
    """
    return int(struct.unpack("!H", buffer)[0])

def deserialize_u32(buffer):
    """
    Unpacks buffer into a uint32 from network order (big endian into the machine endianness)
    """
    return int(struct.unpack("!L", buffer)[0])

def deserialize_string(buffer, string_length):
    """
    Unpacks buffer of `string_length` bytes into a utf8 string
    """
    return struct.unpack(f"{string_length}s", buffer)[0].decode()

File: client/common/deserialize_utils.go
package common

import "encoding/binary"

/*
	Deserializes an unsigned byte
*/
func DeserializeU8(buffer []byte) uint8 {
	return uint8(buffer[0])
}

/*
	Deserializes buffer into an unsigned short in big endian ordering
*/
func DeserializeU16(buffer []byte) uint16 {
	return binary.BigEndian.Uint16(buffer)
}

/*
	Deserializes buffer into an unsigned int in big endian ordering
*/
func DeserializeU32(buffer []byte) uint32 {
	return binary.BigEndian.Uint32(buffer)
}

/*
	Deserializes byte buffer into a string (the buffer may encode a utf-8 string)
*/
func DeserializeString(buffer []byte) string {
	return string(buffer)
}
